src/System.Private.CoreLib/shared/System/Number.Dragon4.cs

```diff
@@ -384,24 +384,27 @@ private static unsafe uint Dragon4(ulong mantissa, int exponent, uint mantissaHi
             }
             else
             {
-                // In the scenario where the first significand digit is after the cutoff, we want to treat that
-                // first significand digit as the rounding digit and increase the decimalExponent by one. This
-                // ensures we correctly handle the case where the first significand digit is exactly one after
+                // In the scenario where the first significant digit is after the cutoff, we want to treat that
+                // first significant digit as the rounding digit. If the first significant would cause the next
+                // digit to round, we will increase the decimalExponent by one and set the previous digit to one.
+                // This  ensures we correctly handle the case where the first significant digit is exactly one after
                 // the cutoff, it is a 4, and the subsequent digit would round that to 5 inducing a double rounding
-                // bug when NumberToString is does its own rounding checks.
-
-                decimalExponent++;
+                // bug when NumberToString does its own rounding checks. However, if the first significant digit
+                // would not cause the next one to round, we preserve that digit as is.
 
                 // divide out the scale to extract the digit
                 outputDigit = BigInteger.HeuristicDivide(ref scaledValue, ref scale);
                 Debug.Assert(outputDigit < 10);
 
                 if ((outputDigit > 5) || ((outputDigit == 5) && !scaledValue.IsZero()))
                 {
-                    buffer[curDigit] = (byte)('1');
-                    curDigit += 1;
+                    decimalExponent++;
+                    outputDigit = 1;
                 }
 
+                buffer[curDigit] = (byte)('0' + outputDigit);
```

Comment:
Here, we either have `1` or the original digit. If the original digit is `5`, then the trailing digits were all zero.

This ensures we are not incorrectly reporting `0`, when we actually had a value that would have parsed back to `double.Epsilon` (or `float.Epsilon`).